coastguard/coastguard.md

```diff
@@ -0,0 +1,138 @@
+
+Creating the local network policy is done in an immutable way, and if in a given timeframe no changes occurred, the controller will simply apply the same local network policy to the k8s api server. 
+
+As part of the project there is a need to examine the interval for which the reconciliation loop should run. It is necessary to find the threshold that will meet the required performance in the project, as this interval represents the biggest trade off in this proposal.
+
+
+In our proposal we currently ignore:
+1. implementation for different CNIs and different network plugins (Calico, Cilium, etc.).
+2. Authentication of network policies or remote pods based on tokens or certificates.
+ 
+
+
+## Design Details
+The following is a diagram of the main “reconciliation loop”. This is our main flow, and any subsequent will use this flow to do the bulk of the work.
+
+![image](https://user-images.githubusercontent.com/42589698/146993706-c19bdfa6-be3a-486e-8d46-fb875f169d78.png)
+
+We will have 4 main flows:
+  1.	a clusters joins the cluster set 
+  2.	a deployment change happens on any cluster  
+  3.	a user adds/changes new multi-network policy is added
+  4.	a cluster is removed from the set 
+
+Flows 2-3 have no special additions beyond executing the reconciliation loop.
+For flow 1 - when a cluster first joins the cluster set, and the coastguard controller runs for the first time, the controller will need to map all existing namespaces and pods on the k8s cluster. It will do so by running the API equivalent of `kubectl get namespaces` and `kubectl get pods`.
+
+For flow 4 - once a cluster leaves the cluster set, the coastguard controller on the leaving cluster should delete all it’s entries from etcd. 
+
+## Component/Service Coast guard controller
+
+1. ETCD structure
+In general, etcd is a key value store. It has the ability to query based on key prefixes.
+So, for each pod, we will have the following entries, where `$clustername` for example is a placeholder for clustername.
+1.	`$clustername_$namespace_$podname`: ip address
+2.	`label_$label_$clustername_$namespace_$podname`: ip address 
+
+We will need the following queries:
+1.	get all pod ip, in cluster X with namespace Y
+  query: etcdctl get --prefix `$clustername_$namespace`
+2.	get all pod ips  in cluster X with namespace Y with label Z
+  query:  etcdctl get --prefix `label_$labelZ_$clusternameX_$namespaceY`
+
+3.	get all pod ips  in cluster with label Z
+  query:  etcdctl get --prefix `label_$labelZ_$clusternameX`
+  Same idea as 2.
+
+#note: ETCD is holding all the sensitive data in our network, the ETCD security is out of scoop in this project.
+2.	CRD schema
+our multi-cluster network policy CRD:
```

Comment:
Is this an example or a definition?

In either case it's not detailed enough, but I'd expect this API to be the main point of this proposal since we need to understand **what** we're trying to achieve before trying to understand **how** we will achieve that.